doc/developer/design/20230728_left_join_stack_consolidation.md

```diff
@@ -0,0 +1,160 @@
+
+Note that the problem occurs regardless of whether SemijoinIdempotence kicks in.
+
+# Explanation
+[explanation]: #explanation
+
+As mentioned above, our proposed solution is to simply perform a consolidation immediately after each of the Union-Negates. More specifically, during the MIR-to-LIR lowering of a Union, we'll check if the Union has an input that has a Negate at the top, and if yes, then we set a `consolidate` flag on the Union, which will make the rendering insert a `consolidate` call on the Union's result. This will immediately make the negated records cancel out.
+
+## Extra memory usage
+
+How `consolidate` works is that it buffers up some data before it actually consolidates. In steady state this won't result in much extra memory usage, because the consolidation happens inside a single timestamp. When processing a snapshot, we will buffer up more data temporarily. This amount is bounded by the snapshot size in the worst case, but is often less, due to compacting in power-of-2-sized chunks already as the data is coming in. (Although, the two inputs of the Union might come in one after the other (due to one having an extra join compared to the other), so we might need to reach large powers of 2 before some consolidation can happen.) Note that if there are several LEFT JOINs, then only two of these `consolidate`s will simultaneously take up memory, because they are pipeline breakers.
+
+The total effect of the feature on memory usage will still be beneficial, as the unconsolidated records are currently consolidated several times when they go into joins (e.g., the 2 unconsolidated records coming from `l1` are currently consolidated both when forming the arrangement for the join in `l2` and also similarly in `l4`). One situation when we might have a net increase in memory usage is when there is only one LEFT JOIN, and there are no other operators downstream that would benefit from a consolidation, but this is probably not so common. Another situation when the consolidation is not helping much is if most records from the left side of a LEFT JOIN don't have a match on the right side.
+
+# Testing and Rollout
+[rollout]: #rollout
+
+We should add tests that check that the physical plan has the `consolidate` flag on Unions as we expect. We should also look at the hierarchical memory visualizer to confirm that record counts decrease as expected. I open to suggestions how to test record counts in an automated way, but I'm afraid that such a test would be too brittle.
+
+We will also add a feature flag for the optimization as a precaution, but it will be enabled by default.
+
```

Comment:
Agreed with your suggestions for using a preventive feature flag, even if turned on by default. Regarding testing, some simple testing will probably be manageable with `testdrive` (since you will need the query repetition feature to get consistent results from queries to these `mz_internal` relations). I would suggest focusing on a join pattern that is cardinality bounding (as usual with fact-to-dimensions), focusing on inserts to get a consistent output number, and getting total message counts (as opposed of breaking down per worker) of chosen operators.

Reply:
Ok, I'll try to create such a test. I've adjusted the text a bit.